### PR TITLE
BackgroundJob to notify admins about app updates

### LINKED ISSUE #3: Backgroundjob to look for updates and send notification to admin

### PR DESCRIPTION
**Pull request by VicDeo**
Closes #3

### CODE REVIEW

File: lib/CheckUpdateBackgroundJob.php [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	 * @param IGroupManager $groupManager
+	 * @param MarketService $marketService
+	 */
+	public function __construct(IConfig $config,
+								ITimeFactory $timeFactory,
+								IManager $notificationManager,
+								IGroupManager $groupManager,
+								MarketService $marketService) {
+		// Run all 10 minutes
+		//$this->setInterval(60 * 10);
```
Comment (VicDeo) [inline lib/CheckUpdateBackgroundJob.php]: set to 24h?

File: lib/CheckUpdateBackgroundJob.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		$this->setInterval(20);
+		$this->config = $config;
+		$this->notificationManager = $notificationManager;
+		$this->groupManager = $groupManager;
+		$this->timeFactory = $timeFactory;
```
Comment (VicDeo) [inline lib/CheckUpdateBackgroundJob.php]: unused

File: lib/CheckUpdateBackgroundJob.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	protected function createNotifications($app, $version, $url) {
+		$lastNotification = $this->config->getAppValue('market', $app, false);
+		if ($lastNotification === $version) {
+			// We already notified about this update
+			return;
+		} else if ($lastNotification !== false) {
+			// Delete old updates
+			$this->deleteOutdatedNotifications($app, $lastNotification);
+		}
+		var_dump($this->notificationManager->listNotifiers());
```
Comment (VicDeo) [inline lib/CheckUpdateBackgroundJob.php]: debug

File: lib/CheckUpdateBackgroundJob.php [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		$this->marketService = $marketService;
+	}
+
+	/**
+	 * @param $argument
+	 */
+	protected function run($argument) {
+		$updates = $this->marketService->getUpdates();
+		foreach ($updates as $appId => $appInfo) {
+			$this->createNotifications($appId, $appInfo['version'], 'ohoho');
```
Comment (VicDeo) [inline lib/CheckUpdateBackgroundJob.php]: Real url to the app description

File: lib/Notifier.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	/**
+	 * Notifier constructor.
+	 *
+	 * @param IManager $notificationManager
+	 * @param IFactory $l10NFactory
+	 */
+	public function __construct(IManager $notificationManager, IFactory $l10NFactory) {
+		$this->notificationManager = $notificationManager;
+		$this->l10NFactory = $l10NFactory;
+		$this->appVersions = $this->getAppVersions();
```
Comment (DeepDiver1975) [inline lib/Notifier.php]: postpone to a later step to save cpu cylces?

File: lib/Notifier.php [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	protected function getCoreVersions() {
+		return implode('.', \OCP\Util::getVersion());
+	}
+
+	protected function getAppVersions() {
+		return \OC_App::getAppVersions();
+	}
+
+	protected function getAppInfo($appId) {
+		return \OC_App::getAppInfo($appId);
```
Comment (DeepDiver1975) [inline lib/Notifier.php]: you can use IAppManager here

File: appinfo/info.xml [hunk 7]
```diff
@@ -9,15 +9,15 @@ Please note: Since ownCloud X (10.0) every instance gets shipped with this app i
 To use this application click on "Files" in the top left corner and click on "Market" (cart icon) (Administrator privileges required) </description>
 	<licence>AGPL</licence>
 	<author>Thomas Müller, Felix Heidecke, Thomas Börger</author>
-	<version>0.2.0</version>
+	<version>0.2.1</version>
```
Comment (DeepDiver1975) [inline appinfo/info.xml]: lets not increment the version for now - lets wait until we do a real release - thx

File: appinfo/info.xml [hunk 8]
```diff
@@ -9,15 +9,15 @@ Please note: Since ownCloud X (10.0) every instance gets shipped with this app i
...
 	<default_enable/>
 	<category>tools</category>
 	<screenshot>https://raw.githubusercontent.com/owncloud/screenshots/master/market/ownCloud-market-app.jpg</screenshot>
 	<dependencies>
 		<owncloud min-version="10.0" max-version="10.0" />
 	</dependencies>
 	<background-jobs>
-		<job>OCA\Market\SyncJob</job>
```
Comment (DeepDiver1975) [inline appinfo/info.xml]: what about the syncjob?

File: lib/Application.php [hunk 9]
```diff
@@ -55,5 +56,19 @@ function ($event) use ($listener) {
 				}
 			}
 		);
+
+		$manager = \OC::$server->getNotificationManager();
```
Comment (DeepDiver1975) [inline lib/Application.php]: @PVince81 we shall come up with a xml based approach for setting up notifications ...

File: appinfo/info.xml (hunk 8 above)
Comment (VicDeo) [inline appinfo/info.xml]: @DeepDiver1975 Somebody blindly copied this from another app. It never existed in market app before. So I think syncjob is worth dying.

File: appinfo/info.xml (hunk 7 above)
Comment (VicDeo) [inline appinfo/info.xml]: @DeepDiver1975 it is needed to register a cronjob properly

Comment (DeepDiver1975) [inline appinfo/info.xml]: very true ..... okay lets keep it then

**Comment by DeepDiver1975** [comment]
bildschirmfoto von 2017-07-28 14-26-47

**Comment by DeepDiver1975** [comment]
and please add unit test - as ususal - thx

**Comment by DeepDiver1975** [comment]
still no unit tests ....

**Comment by DeepDiver1975** [comment]
Tested - works nicely! THX

**Comment by PVince81** [comment]
@VicDeo please backport to stable10 as market app will always be bunded with the tarball and the release script uses stable10.

**Comment by VicDeo** [comment]
Stable10: https://github.com/owncloud/market/pull/128

**Pull request merged**